### PR TITLE
[FLINK-21926][doc] Add docs for fine-grained resource management

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: docs/content/docs/deployment/finegrained_resource.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+## How it works
+
+As described in [Flink Architecture]({{< ref "docs/concepts/flink-architecture" >}}#anatomy-of-a-flink-cluster),
+the resource for task execution in TaskManager is split into a bunch of slots, where job tasks are scheduled.
+The slot is the basic unit of both resource scheduling and resource requirement in Flink runtime.
+
+{{< img src="/fig/dynamic_slot_alloc.png" class="center" >}}
+
+Previously in Flink, the resource requirement only contained the number of the required slots, without fine-grained resource
+profiles, namely **coarse-grained resource management**. The TaskManager contained a fixed number of identical slots to fulfill those requirements.
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:58]:
```suggestion
```

I moved this down

File: docs/content/docs/deployment/finegrained_resource.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+In fine-grained resource management, the slots are requested with specific resource profiles, which can be specified by users.
+Flink will respect those user-specified resource requirements and dynamically cut an exactly-matched slot out of the TaskManager’s available
+resources. As shown above, there is a requirement for a slot with 0.25 Core and 1GB memory and Flink allocates *Slot 1* for it.
+
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:63]:
```suggestion
{{< hint info >}}
Previously in Flink, the resource requirement only contained the number of the required slots, without fine-grained resource
profiles, namely **coarse-grained resource management**. The TaskManager contained a fixed number of identical slots to fulfill those requirements.
{{< /hint >}}

File: docs/content/docs/deployment/finegrained_resource.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+.map(...).slot_sharing_group(ssg_b) # Directly set the slot sharing group with name and resource.
+
+env.register_slot_sharing_group(ssg_a) # Then register the resource of group "a"
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+<div class="alert alert-info">
+     <strong>Note:</strong> Each slot sharing group can only attach to one specified resource, any conflict will fail the compiling of your job.
+</div>
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:177]:
```suggestion
{{< hint warning >}}
**Note:** Each slot sharing group can only attach to one specified resource, any conflict will fail the compiling of your job.
{{< /hint >}}
```

File: docs/content/docs/deployment/finegrained_resource.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+{{< img src="/fig/dynamic_slot_alloc.png" class="center" >}}
+
+Previously in Flink, the resource requirement only contained the number of the required slots, without fine-grained resource
+profiles, namely **coarse-grained resource management**. The TaskManager contained a fixed number of identical slots to fulfill those requirements.
+
+In fine-grained resource management, the slots are requested with specific resource profiles, which can be specified by users.
+Flink will respect those user-specified resource requirements and dynamically cut an exactly-matched slot out of the TaskManager’s available
+resources. As shown above, there is a requirement for a slot with 0.25 Core and 1GB memory and Flink allocates *Slot 1* for it.
+
+For the resource requirement without a specified resource profile, Flink will automatically decide the resource profile of it.
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:64]:
```suggestion
For the resource requirement without a specified resource profile, Flink will automatically decide a resource profile.
```

File: docs/content/docs/deployment/finegrained_resource.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+like GPU are involved, such waste can become even harder to afford. The fine-grained resource management leverages slots of different resources
+to improve resource utilization in such scenarios.
+
+### Resource Allocation Strategy
+
+In this section, we talk about the slot partitioning mechanism in Flink runtime and the resource allocation strategy, including how
+the Flink runtime selects a TaskManager to cut slots and allocates TaskManagers on [Native Kubernetes]({{< ref "docs/deployment/resource-providers/native_kubernetes" >}})
+and [YARN]({{< ref "docs/deployment/resource-providers/yarn" >}}). Note that the resource allocation strategy is pluggable in
+Flink runtime and here we introduce its default implementation in the first step of fine-grained resource
+management. In the future, there might be various strategies that can be selected for different scenarios.
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:300]:
```suggestion
management. In the future, there might be various strategies that users can select for different scenarios.
```

File: docs/content/docs/deployment/finegrained_resource.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Fine-Grained Resource Management
+
+Apache Flink allows you to control the resource consumption of your workload in a finer granularity, namely **fine-grained resource management**.
+It provides means for users to further improve Flink’s resource efficiency with knowledge of their specific scenarios.
```
Comment (sjwiesman) [inline docs/content/docs/deployment/finegrained_resource.md:29]:
```suggestion
Apache Flink works hard to auto-derive sensible default resource requirements for all applications out of the box. 
For users who wish to fine-tune their resource consumption, based on knowledge of their specific scenarios, Flink offers **fine-grained resource management**.
```

I want to make it clear to new users that you don't have to do this.

**Review by sjwiesman — CHANGES_REQUESTED** [review]
I've made a first pass

File: docs/content/docs/deployment/finegrained_resource.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+  - Batch jobs where resources needed for tasks of different stages are significantly different
+
+An in-depth discussion on why fine-grained resource management can improve resource efficiency for the above scenarios is presented in [How it improves resource efficiency](#how-it-improves-resource-efficiency).
+
+## How it works
+
+As described in [Flink Architecture]({{< ref "docs/concepts/flink-architecture" >}}#anatomy-of-a-flink-cluster),
+the resource for task execution in TaskManager is split into a bunch of slots, where job tasks are scheduled.
+The slot is the basic unit of both resource scheduling and resource requirement in Flink runtime.
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply docs/content/docs/deployment/finegrained_resource.md:53]: Maybe `Flink's`?

**Comment by KarmaGYZ** [comment]
Thanks for the valuable comments! @sjwiesman PR updated.

**Comment by KarmaGYZ** [comment]
@sjwiesman Hi, would you like to give it another pass?

**Review by sjwiesman — APPROVED** [review]

**Pull request closed without merging**